### ISSUE TITLE
unable to toggle into full screen using this command python3 -m pygame.examples.aliens

### ISSUE THREAD

**Issue by MattSoo** [pygame/pygame#1952]
pygame 2.0.0.dev6 (SDL 2.0.10, python 3.8.2)
Hello from the pygame community. https://www.pygame.org/contribute.html
Changing to FULLSCREEN
Traceback (most recent call last):
  File "/usr/lib/python3.8/runpy.py", line 193, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.8/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/home/matthew/.local/lib/python3.8/site-packages/pygame/examples/aliens.py", line 411, in <module>
    main()
  File "/home/matthew/.local/lib/python3.8/site-packages/pygame/examples/aliens.py", line 340, in main
    pg.display.flip()
pygame.error: Window surface is invalid, please call SDL_GetWindowSurface() to get a new surface

**Comment by MightyJosip** [comment]
Try to install pygame with `pip install pygame==2.0.0.dev10`

**Comment by MattSoo** [comment]
> Try to install pygame with `pip install pygame==2.0.0.dev10`

It still doesn't work
Changing to FULLSCREEN
ALSA lib pcm.c:8526:(snd_pcm_recover) underrun occurred
ALSA lib pcm.c:8526:(snd_pcm_recover) underrun occurred
X Error of failed request:  BadValue (integer parameter out of range for operation)
  Major opcode of failed request:  153 (XFree86-VidModeExtension)
  Minor opcode of failed request:  10 (XF86VidModeSwitchToMode)
  Value in failed request:  0x3c00013
  Serial number of failed request:  1645
  Current serial number in output stream:  1648

**Comment by MyreMylar** [comment]
What platform is this (Windows, mac, linux etc) ?

What exactly are you doing to change the aliens example to full screen?

**Comment by MattSoo** [comment]
> What platform is this (Windows, mac, linux etc) ?
> 
> What exactly are you doing to change the aliens example to full screen?



> What platform is this (Windows, mac, linux etc) ?
> 
> What exactly are you doing to change the aliens example to full screen?

Linux, Ubuntu to be precise.

**Comment by MyreMylar** [comment]
I assume you are pressing 'f' on they keyboard to switch to full screen in the example.

This is the relevant code in the example. It uses `set_mode()`:

https://github.com/pygame/pygame/blob/7d3900d5b04dc8102f218cc60f3fbf5e06dd8fa1/examples/aliens.py#L323-L340

**Comment by MyreMylar** [comment]
It would probably be useful to print out all the inputs going in to the call to `set_mode()` here.

Changing this:

```
print("Changing to FULLSCREEN") 
screen_backup = screen.copy() 
screen = pg.display.set_mode(SCREENRECT.size, winstyle | pg.FULLSCREEN, bestdepth) 
screen.blit(screen_backup, (0, 0)) 
```

To:

```
print("Changing to FULLSCREEN") 
screen_backup = screen.copy()
print("SCREENRECT.size:", SCREENRECT.size, ", winstyle | pg.FULLSCREEN:", winstyle | pg.FULLSCREEN, ", bestdepth:", bestdepth)
screen = pg.display.set_mode(SCREENRECT.size, winstyle | pg.FULLSCREEN, bestdepth) 
screen.blit(screen_backup, (0, 0)) 
```

Might give more insight as to what is going on.

**Comment by MattSoo** [comment]
> It would probably also be useful to print out all the inputs going in to the call to `set_mode()` here.
> 
> Changing this:
> 
> ```
> print("Changing to FULLSCREEN") 
> screen_backup = screen.copy() 
> screen = pg.display.set_mode(SCREENRECT.size, winstyle | pg.FULLSCREEN, bestdepth) 
> screen.blit(screen_backup, (0, 0)) 
> ```
> 
> To:
> 
> ```
> print("Changing to FULLSCREEN") 
> screen_backup = screen.copy()
> print("SCREENRECT.size:", SCREENRECT.size, ", winstyle | pg.FULLSCREEN:", winstyle | pg.FULLSCREEN, ", bestdepth:", bestdepth)
> screen = pg.display.set_mode(SCREENRECT.size, winstyle | pg.FULLSCREEN, bestdepth) 
> screen.blit(screen_backup, (0, 0)) 
> ```
> 
> Might give more insight as to what is going on.

How can I change it?

**Comment by MyreMylar** [comment]
> How can I change it?

You just need to edit the file `aliens.py` in the `examples` subfolder. These lines:

https://github.com/pygame/pygame/blob/7d3900d5b04dc8102f218cc60f3fbf5e06dd8fa1/examples/aliens.py#L326-L331

If you've downloaded/checked out pygame from github it will be in whichever folder you downloaded it to. If you've installed pygame with pip then it will be in your python directory. For me on windows for 3.8 it's:

`C:\Users\user_name_here\AppData\Local\Programs\Python\Python38-32\Lib\site-packages\pygame\examples`

But it is different on each platform.

**Comment by MyreMylar** [comment]
This doesn't happen on windows. I am able to toggle to fullscreen with the 'F' key in the aliens example (though you have to be quick because it's easy to die which ends the example).

**Comment by MyreMylar** [comment]
This just needs testing on Ubuntu linux to see if it happens there, otherwise we can close as we have no other usable details.

**Comment by siggisv** [comment]
I just tested this on Debian Buster running the Xfce desktop environment (pygame 2.0.0.dev15, SDL 2.0.9 python 3.7.3). Toggling to fullscreen worked fine there.

Tested again on Linux Mint 19.3 running Cinnamon. Toggling to fullscreen worked there in both pygame 2.0.0.dev14 (SDL 2.0.12, python 3.6.9) and pygame 2.0.0.dev6 (SDL 2.0.10 python 3.6.9).

Unless somebody else beats me to it, I could in the next few days install Ubuntu 20.04 on another partition to check if I can manage to trigger the error.

**Note:** Worked in the sense that the game continued running without error messages. The hardware I am running Linux Mint on seems to not support the resolution of the game (640x480) so instead of a fullscreen, I got a borderless window in the top left corner of the screen running the game with some screen artifacts. But I guess that's a different issue.

**Comment by siggisv** [comment]
Bad news.

_... or good news depending on how you view the news that a bug is real **and** we have managed to trigger it._

Just tried downloading virtualbox image of Ubuntu 20.04 (since I couldn't manage to quickly find the download for Ubuntu 20.04, **not** 20.04.1). Both `pygame 2.0.0.dev6` and `pygame 2.0.0.dev14` give the same result when trying to toggle fullscreen when playing `pygame.examples.aliens`.

After adding that single print line suggested above:

```
(venv-pygame2.dev6-testing) osboxes@osboxes:~/work$ python -m pygame.examples.aliens
pygame 2.0.0.dev6 (SDL 2.0.10, python 3.8.5)
Hello from the pygame community. https://www.pygame.org/contribute.html
ALSA lib pcm.c:8526:(snd_pcm_recover) underrun occurred
ALSA lib pcm.c:8526:(snd_pcm_recover) underrun occurred
ALSA lib pcm.c:8526:(snd_pcm_recover) underrun occurred
ALSA lib pcm.c:8526:(snd_pcm_recover) underrun occurred
ALSA lib pcm.c:8526:(snd_pcm_recover) underrun occurred
Changing to FULLSCREEN
SCREENRECT.size: (640, 480) , winstyle | pg.FULLSCREEN: -2147483648 , bestdepth: 24
X Error of failed request:  BadValue (integer parameter out of range for operation)
  Major opcode of failed request:  154 (XFree86-VidModeExtension)
  Minor opcode of failed request:  10 (XF86VidModeSwitchToMode)
  Value in failed request:  0x2e00016
  Serial number of failed request:  473
  Current serial number in output stream:  476
```

```
(venv-pygame2.dev14-testing) osboxes@osboxes:~/work$ python -m pygame.examples.aliens
pygame 2.0.0.dev14 (SDL 2.0.12, python 3.8.5)
Hello from the pygame community. https://www.pygame.org/contribute.html
ALSA lib pcm.c:8526:(snd_pcm_recover) underrun occurred
ALSA lib pcm.c:8526:(snd_pcm_recover) underrun occurred
ALSA lib pcm.c:8526:(snd_pcm_recover) underrun occurred
Changing to FULLSCREEN
SCREENRECT.size: (640, 480) , winstyle | pg.FULLSCREEN: -2147483648 , bestdepth: 24
X Error of failed request:  BadValue (integer parameter out of range for operation)
  Major opcode of failed request:  154 (XFree86-VidModeExtension)
  Minor opcode of failed request:  10 (XF86VidModeSwitchToMode)
  Value in failed request:  0x2e00016
  Serial number of failed request:  464
  Current serial number in output stream:  467
```

**Comment by siggisv** [comment]
Tried to upgrade to 20.04.1 with all current updates and the only difference is that the value in failed request changed:

```
  Value in failed request:  0x1800016
```
Rebooted, value in failed request changed again:

```
  Value in failed request:  0x2400016
```

**Comment by MyreMylar** [comment]
Looking at the internet this error seems to often get triggered when you try to change to an unsupported display mode.

See: https://bbs.archlinux.org/viewtopic.php?id=237080

**Comment by siggisv** [comment]
Further notes from some more tests.

When running `aliens.py` with pygame (tested on both 2.0.0.dev14 and 2.0.0.dev18) installed on Ubuntu 20.04 (or 20.04.1) through `pip install pygame==2.0.0.dev(14 or 18)`:

 - `aliens.py` did manage to change the screen resolution just before crashing with that error message (even though 640x480 wasn't on the list of resolution that the system's display manager offered).
 - When changing the code of `aliens.py` so that it uses 800x600 (an supported resolution according to the system's display manager) then it still crashes with the same error message (after changing the screen resolution).
 - When changing the screen resolution to the same resolution as `aliens.py` before running it, then it does not crash but we get a similar behavior as described in #2118 (sound still playing, but visuals not working: Sometimes it's just not displaying anything at all, sometimes I get a window with a black background (and the area where sprites are drawn onto it gets refreshed), sometimes I get a black window with nothing on it, etc.).
 - When changing the code of `aliens.py` to use `display.toogle_fullscreen()` instead of `display.set_mode()` we still get the same behavior as before.

When running `aliens.py` on Ubuntu 20.04 with a locally compiled pygame (2.0.0.dev19):
- Using the unedited version of `aliens.py` (that uses `display.set_mode()`):
  - Toggling fullscreen on functions perfectly (even when display is not already in the desired resolution).
  - Toggling back to windowed mode changes the resolution but the window it gives me just shows the game as it was when toggling without updating anything (sound still playing, keyboard still working and toggling back to fullscreen still works).
- Using the edited version of `aliens.py` (that uses `display.toggle_fullscreen()`):
  - Toggling between windowed and fullscreen seems to work perfectly, both ways (I did manage **once** to get it to misbehave but have not been successful at repeating that).

**Comment by illume** [comment]
@siggisv does `pygame==1.9.6` installed from wheels work with fullscreen?

I'm trying to figure out if this is a regression or not.

**Comment by siggisv** [comment]
At first, running `pip install pygame==1.9.6` gives error about missing `sdl-config`:

```
  Downloading pygame-1.9.6.tar.gz (3.2 MB)
     |████████████████████████████████| 3.2 MB 7.1 MB/s 
    ERROR: Command errored out with exit status 1:
     command: /home/osboxes/work/venv-pygametest/bin/python3 -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-hdu_osqp/pygame/setup.py'"'"'; __file__='"'"'/tmp/pip-install-hdu_osqp/pygame/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' egg_info --egg-base /tmp/pip-install-hdu_osqp/pygame/pip-egg-info
         cwd: /tmp/pip-install-hdu_osqp/pygame/
    Complete output (12 lines):
    
    
    WARNING, No "Setup" File Exists, Running "buildconfig/config.py"
    Using UNIX configuration...
    
    /bin/sh: 1: sdl-config: not found
    /bin/sh: 1: sdl-config: not found
    /bin/sh: 1: sdl-config: not found
    
    Hunting dependencies...
    WARNING: "sdl-config" failed!
    Unable to run "sdl-config". Please make sure a development version of SDL is installed.
    ----------------------------------------
ERROR: Command errored out with exit status 1: python setup.py egg_info Check the logs for full command output.
```
After installing all dependencies of pygame 1.x.x:

```
(venv-pygametest) osboxes@osboxes:~/work$ pip install pygame==1.9.6
Collecting pygame==1.9.6
  Using cached pygame-1.9.6.tar.gz (3.2 MB)
Building wheels for collected packages: pygame
  Building wheel for pygame (setup.py) ... error
  ERROR: Command errored out with exit status 1:
   command: /home/osboxes/work/venv-pygametest/bin/python3 -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-7jnpy4un/pygame/setup.py'"'"'; __file__='"'"'/tmp/pip-install-7jnpy4un/pygame/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d /tmp/pip-wheel-la5ro122
       cwd: /tmp/pip-install-7jnpy4un/pygame/
  Complete output (12 lines):
  ---
  For help with compilation see:
      https://www.pygame.org/wiki/Compilation
  To contribute to pygame development see:
      https://www.pygame.org/contribute.html
  ---
  usage: setup.py [global_opts] cmd1 [cmd1_opts] [cmd2 [cmd2_opts] ...]
     or: setup.py --help [cmd1 cmd2 ...]
     or: setup.py --help-commands
     or: setup.py cmd --help
  
  error: invalid command 'bdist_wheel'
  ----------------------------------------
  ERROR: Failed building wheel for pygame
  Running setup.py clean for pygame
Failed to build pygame
Installing collected packages: pygame
  Attempting uninstall: pygame
    Found existing installation: pygame 2.0.0.dev18
    Uninstalling pygame-2.0.0.dev18:
      Successfully uninstalled pygame-2.0.0.dev18
    Running setup.py install for pygame ... done
Successfully installed pygame-1.9.6
```

Then running `python -m pygame.examples.aliens` (unedited version):

 - Toggling on fullscreen works.
 - Toggling back to windowed mode crashes with this error message:

```
pygame 1.9.6
Hello from the pygame community. https://www.pygame.org/contribute.html
Changing to FULLSCREEN
Changing to windowed mode
X Error of failed request:  BadValue (integer parameter out of range for operation)
  Major opcode of failed request:  154 (XFree86-VidModeExtension)
  Minor opcode of failed request:  10 (XF86VidModeSwitchToMode)
  Value in failed request:  0x2c00003
  Serial number of failed request:  262
  Current serial number in output stream:  264
```
Running the edited version (the one using `display.toggle_fullscreen()`) gives the same result (toggling to fullscreen works, toggling back to windowed mode crashes with same error message.

**Comment by illume** [comment]
On Ubuntu Bionic with `pygame==2.0.0.dev20` wheel and pygame.examples.aliens: pressing K_f tries to go fullscreen fails, and makes the laptop unresponsive (I had to restart).

**Comment by ObiDriftKenobi** [comment]
This also happens to me. Fullscreen makes the game crash, both 1.9.6 and 2.0.0. I use nvidia proprietary drivers which installed along with the system. Could this be a driver issue?

**Comment by MyreMylar** [comment]
I think we'd narrowed it down to some difference between the linux wheel builds and the dependencies when you build it yourself on ubuntu as full screen seemed to work on this version of ubuntu if built from source, but fail if using the wheel.

**Comment by jrom99** [comment]
I don't know if this helps, but when I tried `python -m pygame.examples.aliens` and tried toggling full screen with F, my Ubuntu 20.10 laptop's screen resolution shrunk from 1366 x 768 to about 400 x 300 while the error was the same (only the last two numbers (576 and 579) change between different tries, but they were always apart by 3 digits)

```
pygame 2.0.0 (SDL 2.0.12, python 3.8.6)
Hello from the pygame community. https://www.pygame.org/contribute.html
Changing to FULLSCREEN
X Error of failed request:  BadValue (integer parameter out of range for operation)
  Major opcode of failed request:  153 (XFree86-VidModeExtension)
  Minor opcode of failed request:  10 (XF86VidModeSwitchToMode)
  Value in failed request:  0x480000e
  Serial number of failed request:  576
  Current serial number in output stream:  579
```

The following code errors out the same way and actually changes my screen's resolution to 800 x 600. If I use another pair of values smaller than my screen size, it chooses the closest available resolution from the options available at Settings > Monitor > Resolution.

- 1024 x 768 -> 1024 x 768
- 1000 x 768 -> 1024 x 768
- 900 x 600 -> 928 x 696

I noticed that sizes bigger than my actual screen work "fine" and create a big white responsive window, but without toggling full-screen. Tried with:
- 6000 x 600
- 600 x 6000
- 6000 x 6000
- 1366 x 6000
- 6000 x 768
- 1366 x 769

```python
import pygame
import sys

pygame.init()
WIDTH, HEIGHT = 800, 600
screen = pygame.display.set_mode((WIDTH, HEIGHT))
pygame.display.toggle_fullscreen()  # offending line

while True:
    pygame.display.update()
    screen.fill((255,255,255))
    for event in pygame.event.get():
        if event.type == pygame.QUIT:
            pygame.quit()
            sys.exit()
        elif event.type == pygame.KEYUP and event.key == pygame.K_ESCAPE:
            pygame.quit()
            sys.exit()
```

However, if I try running it immediately after (without going back to my default resolution) with the default resolution passed as WIDTH, HEIGHT (1366 x 768) the program doesn't crash automatically, but it is just an unresponsive black screen (can't close with escape or using the close button) with size about 1366 x 768.

I tried running the code above with my default resolution first, it showed a semi responsive window that forced me to reboot my computer. That window was flashing white and black and repeatedly moving the cursor to its centre (too fast to allow me to close it or change windows to terminate it). Because of this, I didn't test this code with sizes that would try to change my screen resolution to its default size (e.g.: 1365 x 768 -> 1366 x 768 or something like that).

My computer has "Mesa Intel® HD Graphics 520 (SKL GT2)" in its graphics settings and I'm using X11 instead of Wayland, but I don't know if it is relevant or not to the problem.

**Comment by ankith26** [comment]
This issue should have been fixed in pygame 2.0.3

**Comment by KarsonTheFoxx** [comment]
i get X Error of failed request
